### PR TITLE
Add an option to the example client to simulate WANT_WRITE errors.

### PR DESCRIPTION
**Pull request by haydenroche5**
- Add this option as "-6."
- Turn on non-blocking mode if WANT_WRITE simulation is enabled.
- Create a send IO callback that gets registered when this option is turned on.
  This callback alternates between letting the TX through and returning a
  WANT_WRITE error.

### CODE REVIEW

**Comment by haydenroche5** [comment]
Should we add a unit test to exercise this new feature?

File: examples/client/client.c [hunk 1]
```diff
@@ -1440,6 +1448,49 @@ static void Usage(void)
 #endif
 }
 
+static int SimulateWantWriteIOSendCb(WOLFSSL *ssl, char *buf, int sz, void *ctx)
+{
+    static int wantWriteFlag = 1;
```
Comment (haydenroche5) [inline examples/client/client.c:1453]: This has to be initialized to 1. If 0 instead, it won't exercise WANT_WRITE resilience in the handshake.

File: examples/client/client.c [hunk 2]
```diff
@@ -1440,6 +1448,49 @@ static void Usage(void)
...
+    int sent;
+    int sd = *(int*)ctx;
+
+    (void)ssl;
+
+    if (!wantWriteFlag)
+    {
+        wantWriteFlag = 1;
+
+        sent = wolfIO_Send(sd, buf, sz, 0);
```
Comment (haydenroche5) [inline examples/client/client.c:1464]: Final parameter should be ssl->wflags. Will fix.

File: examples/client/client.c [hunk 3]
```diff
@@ -2110,6 +2162,11 @@ THREAD_RETURN WOLFSSL_THREAD client_test(void* args)
             #endif /* HAVE_TRUSTED_CA */
                 break;
 
+            case '6' :
+                nonBlocking = 1;
```
Comment (haydenroche5) [inline examples/client/client.c:2166]: As written, WANT_WRITE simulation implies non-blocking mode. Should I require the user to specify -N, or leave it like this, and turn on non-blocking mode automatically if the specify -6?

File: examples/server/server.c [hunk 4]
```diff
@@ -307,7 +349,11 @@ static int NonBlockingSSL_Accept(SSL* ssl)
         else
     #endif
         {
-            if (error != WOLFSSL_ERROR_WANT_WRITE) {
+            if (error == WOLFSSL_ERROR_WANT_WRITE)
```
Comment (haydenroche5) [inline examples/server/server.c:352]: Same concept as the changes to the example client.

File: tests/test-tls13.conf [hunk 5]
```diff
@@ -9,10 +9,12 @@
 # server TLSv1.3 TLS13-AES128-GCM-SHA256
 -v 4
 -l TLS13-AES128-GCM-SHA256
+-6
```
Comment (haydenroche5) [inline tests/test-tls13.conf:12]: I just picked one test for TLS 1.3 and one for 1.2, but let me know if I should add more.

File: examples/client/client.c (hunk 3 above)
Comment (toddouska) [reply examples/client/client.c:2163]: Automatically turning on an implied requirement is ok and simplifies help messages in this case.

File: examples/client/client.c [hunk 6]
```diff
@@ -1445,6 +1444,49 @@ static void Usage(void)
 #endif
 }
 
+static int SimulateWantWriteIOSendCb(WOLFSSL *ssl, char *buf, int sz, void *ctx)
```
Comment (toddouska) [inline examples/client/client.c:1447]: Can you move this code in wolfssl/test.h with the other shared callbacks?  That way the server and client don't have to duplicate this code.

Comment (haydenroche5) [reply examples/client/client.c:1447]: Good idea.

**Comment by haydenroche5** [comment]
retest this please

**Review by toddouska — APPROVED** [review]

**Pull request merged**